src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,612 @@
+/*-------------------------------------------------------------------------
+ *
+ * connection_management.c
+ *   Central management of connections and their life-cycle
+ *
+ * Copyright (c) 2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#ifdef HAVE_POLL_H
+#include <poll.h>
+#endif
+
+#include "libpq-fe.h"
+
+#include "miscadmin.h"
+
+#include "access/hash.h"
+#include "commands/dbcommands.h"
+#include "distributed/connection_management.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/hash_helpers.h"
+#include "mb/pg_wchar.h"
+#include "utils/hsearch.h"
+#include "utils/memutils.h"
+
+
+HTAB *ConnectionHash = NULL;
+MemoryContext ConnectionContext = NULL;
+
+static uint32 ConnectionHashHash(const void *key, Size keysize);
+static int ConnectionHashCompare(const void *a, const void *b, Size keysize);
+static MultiConnection * StartConnectionEstablishment(ConnectionHashKey *key);
+static void AfterXactHostConnectionHandling(ConnectionHashEntry *entry, bool isCommit);
+
+
+/*
+ * Initialize per-backend connection management infrastructure.
+ */
+void
+InitializeConnectionManagement(void)
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+
+
+	/*
+	 * Create a single context for connection and transaction related memory
+	 * management. Doing so, instead of allocating in TopMemoryContext, makes
+	 * it easier to associate used memory.
+	 */
+	ConnectionContext = AllocSetContextCreate(TopMemoryContext, "Connection Context",
+											  ALLOCSET_DEFAULT_MINSIZE,
+											  ALLOCSET_DEFAULT_INITSIZE,
+											  ALLOCSET_DEFAULT_MAXSIZE);
+
+	/* create (host,port,user,database) -> [connection] hash */
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(ConnectionHashKey);
+	info.entrysize = sizeof(ConnectionHashEntry);
+	info.hash = ConnectionHashHash;
+	info.match = ConnectionHashCompare;
+	info.hcxt = ConnectionContext;
+	hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT | HASH_COMPARE);
+
+	ConnectionHash = hash_create("citus connection cache (host,port,user,database)",
+								 64, &info, hashFlags);
+}
+
+
+/*
+ * Perform connection management activity after the end of a transaction. Both
+ * COMMIT and ABORT paths are handled here.
+ *
+ * This is called by Citus' global transaction callback.
+ */
+void
+AfterXactConnectionHandling(bool isCommit)
```

Comment:
Not sure what this does from the name or the description. Maybe AfterXactResetConnections?